lib/solvers/HighDensitySolver/SingleHighDensityRouteSolver.ts

```diff
@@ -411,22 +411,24 @@ export class SingleHighDensityRouteSolver extends BaseSolver {
   }
 
   _step() {
-    this.candidates.sort((a, b) => b.f - a.f)
     let currentNode = this.candidates.pop()
+    let currentNodeKey = currentNode ? this.getNodeKey(currentNode) : undefined
 
     while (
       currentNode &&
-      this.exploredNodes.has(this.getNodeKey(currentNode))
+      currentNodeKey &&
+      this.exploredNodes.has(currentNodeKey)
     ) {
       currentNode = this.candidates.pop()
+      currentNodeKey = currentNode ? this.getNodeKey(currentNode) : undefined
     }
 
-    if (!currentNode) {
+    if (!currentNode || !currentNodeKey) {
       this.failed = true
       return
     }
-    this.exploredNodes.add(this.getNodeKey(currentNode))
-    this.debug_exploredNodesOrdered.push(this.getNodeKey(currentNode))
+    this.exploredNodes.add(currentNodeKey)
+    this.debug_exploredNodesOrdered.push(currentNodeKey)
```

Comment:
Small micro optimisation to reduce the number of times the key is generated

Reply:
niceeee